src/main/scala/com/github/catalystcode/fortis/spark/streaming/reddit/client/RedditClient.scala

```diff
@@ -0,0 +1,197 @@
+package com.github.catalystcode.fortis.spark.streaming.reddit.client
+
+import java.net.URLEncoder
+import java.util.UUID
+
+import com.github.catalystcode.fortis.spark.streaming.reddit.Logger
+import com.github.catalystcode.fortis.spark.streaming.reddit.dto.RedditObject
+import net.liftweb.json
+
+import scalaj.http._
+
+/**
+  * The RedditClient class provides two public-facing methods:
+  * <ul>
+  *   <li>{@link #ensureFreshToken()}</li>
+  *   <li>{@link #search()}</li>
+  * </ul>
+  *
+  * @param applicationId
+  * @param secret
+  * @param fetcher
+  */
+@SerialVersionUID(100L)
+class RedditClient(val applicationId: String,
+                   val secret: String,
+                   val fetcher: ResponseFetcher = new DefaultResponseFetcher())
+  extends Serializable with Logger {
+
+  protected val userAgent = s"FotisReddit/0.1 by ${this.applicationId}"
+  protected val deviceId = UUID.randomUUID().toString
+  protected var clientCredentials: Option[ClientCredentials] = None
+
+  /**
+    * Attempts to fetch a fresh token from Reddit. If no valid token is present, a new one is fetched. Otherwise, an
+    * attempt to refresh the existing token is made.
+    */
+  def ensureFreshToken(): Unit = {
+    if (clientCredentials.isEmpty) {
+      fetchAccessToken()
+    }
+    else {
+      try {
+        refreshToken()
+      }
+      catch {
+        case ex: Exception =>
+          logError(s"Exception while attempting to refresh Reddit token", ex)
+          fetchAccessToken()
+      }
+    }
+  }
+
+  def tokenExpirationInSeconds(): Option[Int] = {
+    if (clientCredentials.isDefined) Option(clientCredentials.get.expires_in) else None
```

Comment:
Not that I'm a fan of this, but just to note that this can be simplified using Scala idioms to `clientCredentials.map(_.expires_in)` (maybe it's flatMap, but one of the two should compile). It's up to you to decide which one you find more readable.